File: go/linkedlist/CircularLinkedList/CircularLinkedList_test.go
package CircularLinkedList_test

import (
	"testing"

	linkedlist "example.com/go/go/linkedlist/CircularLinkedList"
	"gotest.tools/v3/assert"
)

func TestCircularLinkedListIsEmpty(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	assert.Equal(t, l.IsEmpty(), true)
}

func TestCircularLinkedListSize(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.Append(1)
	l.Append(2)
	l.Append(3)
	assert.Equal(t, l.Length, 3)
}
func TestCircularLinkedListPrepend(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.Prepend(1)
	assert.Equal(t, l.Head.Value, 1)
	l.Prepend(2)
	assert.Equal(t, l.Head.Value, 2)
	assert.Equal(t, l.IsCircular(), true)
	assert.DeepEqual(t, l.ToArray(), []int{2, 1})

	l.Prepend(3)
	l.Prepend(4)

	assert.DeepEqual(t, l.ToArray(), []int{4, 3, 2, 1})
	assert.DeepEqual(t, l.ReverseToArray(), []int{1, 2, 3, 4})
	assert.Equal(t, l.IsCircular(), true)
}
func TestCircularLinkedListAppend(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.Append(1)
	l.Append(2)
	l.Append(3)
	l.Append(4)
	assert.Equal(t, l.Length, 4)
	assert.Equal(t, l.IsCircular(), true)
	assert.DeepEqual(t, l.ToArray(), []int{1, 2, 3, 4})
	assert.DeepEqual(t, l.ReverseToArray(), []int{4, 3, 2, 1})
}

func TestCircularLinkedListInsertAtStart(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.InsertAt(1, 0)
	l.InsertAt(2, 0)
	l.InsertAt(3, 0)
	assert.Equal(t, l.Length, 3)
	assert.Equal(t, l.IsCircular(), true)
	assert.DeepEqual(t, l.ToArray(), []int{3, 2, 1})
	assert.DeepEqual(t, l.ReverseToArray(), []int{1, 2, 3})
}

func TestCircularLinkedListInsertAtEnd(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.InsertAt(1, 0)
	l.Append(2)
	l.Append(3)
	l.InsertAt(4, 3)
	l.InsertAt(5, 4)
	assert.Equal(t, l.Length, 5)
	assert.Equal(t, l.IsCircular(), true)
	assert.DeepEqual(t, l.ToArray(), []int{1, 2, 3, 4, 5})
	assert.DeepEqual(t, l.ReverseToArray(), []int{5, 4, 3, 2, 1})
}

func TestCircularLinkedListInsertAtMiddle(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.Append(1)
	l.Append(2)
	l.Append(3)
	l.Append(4)

	assert.Equal(t, l.Length, 4)
	l.InsertAt(5, 2)
	assert.Equal(t, l.Length, 5)
	assert.DeepEqual(t, l.ToArray(), []int{1, 2, 5, 3, 4})
	assert.DeepEqual(t, l.ReverseToArray(), []int{4, 3, 5, 2, 1})
	assert.Equal(t, l.IsCircular(), true)

	l.InsertAt(6, 3)
	assert.Equal(t, l.Length, 6)
	assert.DeepEqual(t, l.ToArray(), []int{1, 2, 5, 6, 3, 4})
	assert.DeepEqual(t, l.ReverseToArray(), []int{4, 3, 6, 5, 2, 1})
	assert.Equal(t, l.IsCircular(), true)
}

func TestCircularLinkedListDeleteAt(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	l.Append(1)
	l.Append(2)
	l.Append(3)
	l.Append(4)
	assert.Equal(t, l.Length, 4)
	assert.Equal(t, l.IsCircular(), true)
	assert.DeepEqual(t, l.ToArray(), []int{1, 2, 3, 4})

	l.DeleteAt(0)

	assert.Equal(t, l.Length, 3)
	assert.DeepEqual(t, l.ToArray(), []int{2, 3, 4})
	assert.DeepEqual(t, l.ReverseToArray(), []int{4, 3, 2})
	assert.Equal(t, l.IsCircular(), true)

	l.DeleteAt(2)
	assert.Equal(t, l.Length, 2)
	assert.DeepEqual(t, l.ToArray(), []int{2, 3})
	assert.DeepEqual(t, l.ReverseToArray(), []int{3, 2})
	assert.Equal(t, l.IsCircular(), true)

	l.Append(4)
	l.Append(5)
	l.Append(6)
	l.Append(7)
	assert.Equal(t, l.Length, 6)

	l.DeleteAt(5)
	assert.Equal(t, l.Length, 5)
	assert.DeepEqual(t, l.ToArray(), []int{2, 3, 4, 5, 6})
	assert.DeepEqual(t, l.ReverseToArray(), []int{6, 5, 4, 3, 2})
	assert.Equal(t, l.IsCircular(), true)

	l.DeleteAt(0)
	assert.Equal(t, l.Length, 4)
	assert.DeepEqual(t, l.ToArray(), []int{3, 4, 5, 6})
	assert.DeepEqual(t, l.ReverseToArray(), []int{6, 5, 4, 3})
	assert.Equal(t, l.IsCircular(), true)

	l.DeleteAt(3)
	assert.Equal(t, l.Length, 3)
	assert.DeepEqual(t, l.ToArray(), []int{3, 4, 5})
	assert.DeepEqual(t, l.ReverseToArray(), []int{5, 4, 3})
	assert.Equal(t, l.IsCircular(), true)

	l.DeleteAt(1)
	assert.Equal(t, l.Length, 2)
	assert.DeepEqual(t, l.ToArray(), []int{3, 5})
	assert.DeepEqual(t, l.ReverseToArray(), []int{5, 3})

	assert.Equal(t, l.IsCircular(), true)

	assert.Error(t, l.DeleteAt(2), "index out of range")
	assert.Error(t, l.DeleteAt(-1), "index out of range")
	assert.Error(t, l.DeleteAt(3), "index out of range")
	l.DeleteAt(0)
	l.DeleteAt(0)

	assert.Equal(t, l.Length, 0)
	assert.Equal(t, l.IsCircular(), false)
}

func TestCircularLinkedListError(t *testing.T) {
	l := linkedlist.NewCircularLinkedList()
	assert.Error(t, l.DeleteAt(0), "index out of range")
	assert.Error(t, l.InsertAt(1, 1), "index out of range")
	assert.Error(t, l.InsertAt(1, -1), "index out of range")
}

File: Java/src/test/java/linkedlist/CircularLinkedList/CircularLinkedListTest.java
package linkedlist.CircularLinkedList;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.example.linkedlist.CircularLinkedList.CircularLinkedList;

public class CircularLinkedListTest {

    @Test
    public void testCircularLinkedListIsEmpty() {
        CircularLinkedList list = new CircularLinkedList();
        Assertions.assertTrue(list.isEmpty());
        Assertions.assertEquals(0, list.size());
        list.append(1);
        Assertions.assertFalse(list.isEmpty());
        Assertions.assertEquals(1, list.size());
    }

    @Test
    public void testCircularLinkedListAppend() {
        CircularLinkedList list = new CircularLinkedList();
        list.append(1);
        Assertions.assertEquals(1, list.size());
        Assertions.assertArrayEquals(new int[] { 1 }, list.toArray());
        list.append(2);
        Assertions.assertEquals(2, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2 }, list.toArray());
        list.append(3);
        Assertions.assertEquals(3, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.append(4);
        Assertions.assertEquals(4, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3, 4 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 4, 3, 2, 1 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

    }

    @Test
    public void testCircularLinkedListPrepend() {
        CircularLinkedList list = new CircularLinkedList();
        list.prepend(1);
        Assertions.assertEquals(1, list.size());
        Assertions.assertArrayEquals(new int[] { 1 }, list.toArray());
        list.prepend(2);
        Assertions.assertEquals(2, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 1 }, list.toArray());
        list.prepend(3);
        Assertions.assertEquals(3, list.size());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.prepend(4);
        Assertions.assertEquals(4, list.size());
        Assertions.assertArrayEquals(new int[] { 4, 3, 2, 1 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3, 4 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.prepend(5);
        Assertions.assertEquals(5, list.size());
        Assertions.assertArrayEquals(new int[] { 5, 4, 3, 2, 1 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3, 4, 5 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());
    }

    @Test
    public void testCircularLinkedListInsertAt() {
        CircularLinkedList list = new CircularLinkedList();
        list.insertAt(1, 0);
        Assertions.assertEquals(1, list.size());
        Assertions.assertArrayEquals(new int[] { 1 }, list.toArray());
        list.insertAt(2, 1);
        Assertions.assertEquals(2, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2 }, list.toArray());
        list.insertAt(3, 2);
        Assertions.assertEquals(3, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

        list.insertAt(4, 0);
        Assertions.assertEquals(4, list.size());
        Assertions.assertArrayEquals(new int[] { 4, 1, 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1, 4 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.insertAt(5, 1);
        Assertions.assertEquals(5, list.size());
        Assertions.assertArrayEquals(new int[] { 4, 5, 1, 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1, 5, 4 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

        list.insertAt(6, 2);
        Assertions.assertEquals(6, list.size());
        Assertions.assertArrayEquals(new int[] { 4, 5, 6, 1, 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2, 1, 6, 5, 4 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.insertAt(7, -1);
        });

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.insertAt(7, 7);
        });
    }

    @Test
    public void testCircularLinkedListDeleteAt() {
        CircularLinkedList list = new CircularLinkedList();
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(0);
        });
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        Assertions.assertEquals(4, list.size());
        Assertions.assertArrayEquals(new int[] { 1, 2, 3, 4 }, list.toArray());

        list.deleteAt(0);
        Assertions.assertEquals(3, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 3, 4 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 4, 3, 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.deleteAt(2);
        Assertions.assertEquals(2, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 3 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 3, 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

        list.append(4);
        list.append(8);
        list.append(5);
        Assertions.assertEquals(5, list.size());

        list.deleteAt(3);
        Assertions.assertEquals(4, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 3, 4, 5 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 5, 4, 3, 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(-1);
        });

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(5);
        });

        list.deleteAt(3);
        Assertions.assertEquals(3, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 3, 4 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 4, 3, 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());
        Assertions.assertFalse(list.isEmpty());

        list.deleteAt(1);
        Assertions.assertEquals(2, list.size());
        Assertions.assertArrayEquals(new int[] { 2, 4 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 4, 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

        list.deleteAt(1);
        Assertions.assertEquals(1, list.size());
        Assertions.assertArrayEquals(new int[] { 2 }, list.toArray());
        Assertions.assertArrayEquals(new int[] { 2 }, list.reverseToArray());
        Assertions.assertTrue(list.isCircular());

        list.deleteAt(0);
        Assertions.assertEquals(0, list.size());
        Assertions.assertArrayEquals(new int[] {}, list.toArray());
        Assertions.assertArrayEquals(new int[] {}, list.reverseToArray());
        Assertions.assertFalse(list.isCircular());
        Assertions.assertTrue(list.isEmpty());
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(0);
        });

    }


    @Test
    public void testCircularLinkedListFindAt() {
        CircularLinkedList list = new CircularLinkedList();
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.findAt(0);
        });
        list.append(1);

        Assertions.assertEquals(1, list.findAt(0));

        list.append(2);
        Assertions.assertEquals(2, list.findAt(1));

        list.append(3);
        Assertions.assertEquals(3, list.findAt(2));

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.findAt(-1);
        });
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.findAt(3);
        });
    }
}
